README.md
Joachim Gassen

# Download, Tidy and Visualize Covid-19 Related Data

## Disclaimer

I am an applied economist studying the economic effects of regulatory
interventions on corporate transparency and leading the Open Science
Data Center (OSDC) of the [TRR 266 Accounting for
Transparency](https://accounting-for-transparency.de), which is funded
by the German Science Foundation (DFG). The OSDC has the objective to
make research transparent in a way that others can contribute and
collaborate.

This is the spirit that motivated me to set up this package. I am
clearly no epidemiologist so I will abstain from from providing
infrastructure for analyzing the spread of the disease or estimating the
effects of non-pharmaceutical interventions. Instead this package serves
the purpose to facilitate the use of various Covid-19 related data
sources with a special focus on non-pharmaceutical interventions.

In that way, I hope that it might be helpful for others that are
interested in doing research on the Covid 19 pandemic by promoting the
benefits of open science.

## Why yet another package on Covid-19?

There are several packages that provide data and infrastructure related
to Covid-19. Two prominent cases are:

  - `{nCov2019}`: This
    [package](https://github.com/GuangchuangYu/nCov2019) has a focus on
    Chinese data but also contains data on other countries and regions.
    It contains a shiny dashboard.
  - `{conronavirus}`: This
    [package](https://github.com/RamiKrispin/coronavirus) provides the
    Johns Hopkins University CSSE dataset together with a dashboard

Additional R related resources on Covid-19 can be found
[here](https://www.statsandr.com/blog/top-r-resources-on-covid-19-coronavirus/)
and [here](https://github.com/mine-cetinkaya-rundel/covid19-r).

Other than the packages mentioned above, the key objective of the
{tidycovid19} package is to provide *transparent* access to *various*
data sources at the country-day level, including data on governmental
interventions and on behavioral response of the public. It does not
contain any data per se. Instead, it provides functions to pull data
from publicly available authoritative sources. The sources and the data
are documented by additional data frames included in the package. While
the combined data frame generated by `download_merged_data()` aggregates
data at the country-day level, some functions also provide sub-country
level data on request.

For those interested in speedy downloads it alternatively provides the
option to download from the cached data in this repo (stored in the
directory `cached_data`). The cached data is updated daily.

## The Data

Currently, the package offers the following functions to download data:

  - `download_jhu_csse_covid19_data()`: Downloads and tidies [Covid-19
    data from the Johns Hopkins University CSSE Github
    Repo](https://github.com/CSSEGISandData/COVID-19). This data has
    developed to a standard resource for researchers and the general
    audience interested in assessing the global spreading of the virus.
    The data is provided at country and sub-country levels.
  - `download_ecdc_covid19_data()`: Downloads and tidies [Covid-19 case
    data provided by the European Centre for Disease Prevention and
    Control](https://www.ecdc.europa.eu/en/publications-data/download-todays-data-geographic-distribution-covid-19-cases-worldwide).
    The data is updated daily and contains the latest available public
    data on the number of new Covid-19 cases reported per day and per
    country.
  - **NEW**: `download_owid_data()`: Downloads and tidies [data
    collected by the ‘Our World in Data’
    team](https://ourworldindata.org/covid-testing). This team
    systematically collects data on hospitalizations, testing and
    vaccinations from multiple national sources.
  - `download_acaps_npi_data()`: Downloads and tidies the [Government
    measures dataset provided by the Assessment Capacities Project
    (ACAPS)](https://www.acaps.org/covid19-government-measures-dataset).
    These data allow researchers to study the effect of
    non-pharmaceutical interventions on the development of the virus.
  - `download_oxford_npi_data()`: Downloads and tidies data from the
    [Oxford Covid-19 Government Response
    Tracker](https://www.bsg.ox.ac.uk/research/research-projects/oxford-covid-19-government-response-tracker),
    an alternative data source for governmental interventions.
    Currently, I do not include the financial measures of this data set
    and also do not include its data in the merged data file (see below)
    as I view the ACAPS data to be better suited ([this blog post
    details
    why](https://joachim-gassen.github.io/2020/04/exploring-and-benchmarking-oxford-government-response-data/)).
  - `download_apple_mtr_data()`: Downloads [Mobility Trends Reports
    provided by Apple](https://www.apple.com/covid19/mobility) related
    to Covid-19. The data is provided at country and sub-country levels.
  - `download_google_cmr_data()`: Downloads [Google COVID-19 Community
    Mobility Reports](https://www.google.com/covid19/mobility/) data. As
    of April 17, Google provides a nice and clean CSV file containing
    country-day and region-day data. This makes the PDF scraping code
    that used to be part of this package obsolete. If you are interested
    in it for didactic reasons, you can still find it in the [git
    hstory](https://github.com/joachim-gassen/tidycovid19/tree/947f010bccb111bc181ca31889c6f2f86b841fb5/R).
    This data is available at the country, regional and U.S. county
    level.
  - `download_google_trends_data()`: Downloads and tidies [Google
    Trends](https://trends.google.com/trends/) data on the search volume
    for the term “coronavirus” (Thank you to Yan Ouaknine for bringing
    up that idea\!). This data can be used to assess the public
    attention to Covid-19 across countries and over time within a given
    country. The data is available at the country, regional and city
    level but availability varies across countries as Google Trends
    provides only more granular data for regions with high search
    volumes.
  - `download_wbank_data()`: Downloads and tidies additional country
    level information provided by the [World
    Bank](https://data.worldbank.org) using the {wbstats} package. These
    data allow researchers to calculate per capita measures of the virus
    spread and to assess the association of macro-economic variables
    with the development of the virus.
  - `download_merged_data()`: Downloads all data sources and creates a
    merged country-day panel.

## How to Use the Package

The idea is simple. Load the data using the functions above and code
away. So, for example:

``` r
# Suggestion by AndreaPi (issue #19)

library(tidyverse)
library(tidycovid19)
library(zoo)

df <- download_merged_data(cached = TRUE, silent = TRUE)

df %>%
  filter(iso3c == "USA") %>%
  mutate(
    new_cases = confirmed - lag(confirmed),
    ave_new_cases = rollmean(new_cases, 7, na.pad=TRUE, align="right")
  ) %>%
  filter(!is.na(new_cases), !is.na(ave_new_cases)) %>%
  ggplot(aes(x = date)) +
  geom_bar(aes(y = new_cases), stat = "identity", fill = "lightblue") +
  geom_line(aes(y = ave_new_cases), color ="red") +
  theme_minimal()
```

<img src="man/figures/Example-1.png" style="display: block; margin: auto;" />

The data comes with two meta data sets that describe the data. The data
frame `tidycovid19_data_sources` provides short descriptions and links
for each data source used by the package. The data frame
`tidycovid19_variable_defintions` provides variable definitions for each
variable included in the merged country-day data frame provided by
`download_merged_data()`:

``` r
df <- tidycovid19_variable_definitions %>%
  select(var_name, var_def)
kable(df) %>% kableExtra::kable_styling()
```

<table class="table" style="margin-left: auto; margin-right: auto;">

<thead>

<tr>

<th style="text-align:left;">

var\_name

</th>

<th style="text-align:left;">

var\_def

</th>

</tr>

</thead>

<tbody>

<tr>

<td style="text-align:left;">

iso3c

</td>

<td style="text-align:left;">

ISO3c country code as defined by ISO 3166-1 alpha-3

</td>

</tr>

<tr>

<td style="text-align:left;">

country

</td>

<td style="text-align:left;">

Country name

</td>

</tr>

<tr>

<td style="text-align:left;">

date

</td>

<td style="text-align:left;">

Calendar date

</td>

</tr>

<tr>

<td style="text-align:left;">

confirmed

</td>

<td style="text-align:left;">

Confirmed Covid-19 cases as reported by JHU CSSE (accumulated)

</td>

</tr>

<tr>

<td style="text-align:left;">

deaths

</td>

<td style="text-align:left;">

Covid-19-related deaths as reported by JHU CSSE (accumulated)

</td>

</tr>

<tr>

<td style="text-align:left;">

recovered

</td>

<td style="text-align:left;">

Covid-19 recoveries as reported by JHU CSSE (accumulated)

</td>

</tr>

<tr>

<td style="text-align:left;">

ecdc\_cases

</td>

<td style="text-align:left;">

Covid-19 cases as reported by ECDC (accumulated)

</td>

</tr>

<tr>

<td style="text-align:left;">

ecdc\_deaths

</td>

<td style="text-align:left;">

Covid-19-related deaths as reported by ECDC (accumulated)

</td>

</tr>

<tr>

<td style="text-align:left;">

total\_tests

</td>

<td style="text-align:left;">

Accumulated test counts as reported by Our World in Data

</td>

</tr>

<tr>

<td style="text-align:left;">

tests\_units

</td>

<td style="text-align:left;">

Definition of what constitutes a ‘test’

</td>

</tr>

<tr>

<td style="text-align:left;">

positive\_rate

</td>

<td style="text-align:left;">

The share of COVID-19 tests that are positive, given as a rolling 7-day
average

</td>

</tr>

<tr>

<td style="text-align:left;">

hosp\_patients

</td>

<td style="text-align:left;">

Number of COVID-19 patients in hospital on a given day

</td>

</tr>

<tr>

<td style="text-align:left;">

icu\_patients

</td>

<td style="text-align:left;">

Number of COVID-19 patients in intensive care units (ICUs) on a given
day

</td>

</tr>

<tr>

<td style="text-align:left;">

total\_vaccinations

</td>

<td style="text-align:left;">

Total number of COVID-19 vaccination doses administered

</td>

</tr>

<tr>

<td style="text-align:left;">

soc\_dist

</td>

<td style="text-align:left;">

Number of social distancing measures reported up to date by ACAPS, net
of lifted restrictions

</td>

</tr>

<tr>

<td style="text-align:left;">

mov\_rest

</td>

<td style="text-align:left;">

Number of movement restrictions reported up to date by ACAPS, net of
lifted restrictions

</td>

</tr>

<tr>

<td style="text-align:left;">

pub\_health

</td>

<td style="text-align:left;">

Number of public health measures reported up to date by ACAPS, net of
lifted restrictions

</td>

</tr>

<tr>

<td style="text-align:left;">

gov\_soc\_econ

</td>

<td style="text-align:left;">

Number of social and economic measures reported up to date by ACAPS, net
of lifted restrictions

</td>

</tr>

<tr>

<td style="text-align:left;">

lockdown

</td>

<td style="text-align:left;">

Number of lockdown measures reported up to date by ACAPS, net of lifted
restrictions

</td>

</tr>

<tr>

<td style="text-align:left;">

apple\_mtr\_driving

</td>

<td style="text-align:left;">

Apple Maps usage for driving directions, as percentage\*100 relative to
the baseline of Jan 13, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

apple\_mtr\_walking

</td>

<td style="text-align:left;">

Apple Maps usage for walking directions, as percentage\*100 relative to
the baseline of Jan 13, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

apple\_mtr\_transit

</td>

<td style="text-align:left;">

Apple Maps usage for public transit directions, as percentage\*100
relative to the baseline of Jan 13, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_retail\_recreation

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit retail and recreation places expressed as a percentage\*100 change
relative to the baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_grocery\_pharmacy

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit grocery stores and pharmacies expressed as a percentage\*100
change relative to the baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_parks

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit parks expressed as a percentage\*100 change relative to the
baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_transit\_stations

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit transit stations expressed as a percentage\*100 change relative to
the baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_workplaces

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit workplaces expressed as a percentage\*100 change relative to the
baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gcmr\_residential

</td>

<td style="text-align:left;">

Google Community Mobility Reports data for the frequency that people
visit residential places expressed as a percentage\*100 change relative
to the baseline period Jan 3 - Feb 6, 2020

</td>

</tr>

<tr>

<td style="text-align:left;">

gtrends\_score

</td>

<td style="text-align:left;">

Google search volume for the term ‘coronavirus’, relative across time
with the country maximum scaled to 100

</td>

</tr>

<tr>

<td style="text-align:left;">

gtrends\_country\_score

</td>

<td style="text-align:left;">

Country-level Google search volume for the term ‘coronavirus’ over a
period starting Jan 1, 2020, relative across countries with the country
having the highest search volume scaled to 100 (time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

region

</td>

<td style="text-align:left;">

Country region as classified by the World Bank (time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

income

</td>

<td style="text-align:left;">

Country income group as classified by the World Bank (time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

population

</td>

<td style="text-align:left;">

Country population as reported by the World Bank (original identifier
‘SP.POP.TOTL’, time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

land\_area\_skm

</td>

<td style="text-align:left;">

Country land mass in square kilometers as reported by the World Bank
(original identifier ‘AG.LND.TOTL.K2’, time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

pop\_density

</td>

<td style="text-align:left;">

Country population density as reported by the World Bank (original
identifier ‘EN.POP.DNST’, time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

pop\_largest\_city

</td>

<td style="text-align:left;">

Population in the largest metropolian area of the country as reported by
the World Bank (original identifier ‘EN.URB.LCTY’, time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

life\_expectancy

</td>

<td style="text-align:left;">

Average life expectancy at birth of country citizens in years as
reported by the World Bank (original identifier ‘SP.DYN.LE00.IN’,
time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

gdp\_capita

</td>

<td style="text-align:left;">

Country gross domestic product per capita, measured in 2010 US-$ as
reported by the World Bank (original identifier ‘NY.GDP.PCAP.KD’,
time-stable)

</td>

</tr>

<tr>

<td style="text-align:left;">

timestamp

</td>

<td style="text-align:left;">

Date and time where data has been collected from authoritative sources

</td>

</tr>

</tbody>

</table>

There are more examples on how to code in the code file in the main
directory with the revealing name `code_examples.R`. Explore and reuse\!

## Visualization

The focus of the package lies on data collection and not on
visualization as there are already many great tools floating around.
Regardless, there are three functions that allow you to visualize some
of the key data that the package provides.

### Plot Covid-19 Spread over Event Time

The function `plot_covid19_spread()` allows you to quickly visualize the
spread of the virus in relation to governmental intervention measures.
It is inspired by the insightful displays created by John Burn-Murdoch
from the Financial Times and offers various customization options.

``` r
#remotes::install_github("joachim-gassen/tidycovid19")

library(tidycovid19)

merged <- download_merged_data(cached = TRUE, silent = TRUE)
plot_covid19_spread(
  merged, highlight = c("ITA", "ESP", "GBR", "FRA", "DEU", "USA", "BRA", "MEX"),
  intervention = "lockdown", edate_cutoff = 330
)
```

<img src="man/figures/DemoPlot-1.png" style="display: block; margin: auto;" />

### Plot Covid-19 Stripes

Another option to visualize the spread of Covid-19, in particular if you
want to compare many countries, is to produce a stripes-based
visualization. Meet the Covid-19 stripes:

``` r
plot_covid19_stripes()
```

<img src="man/figures/Covid19Stripes-1.png" style="display: block; margin: auto;" />

Again, the function comes with many options. As an example, you can
easily switch to a per capita display:

``` r
plot_covid19_stripes(
  per_capita = TRUE, 
  population_cutoff = TRUE, 
  sort_countries = "magnitude"
)
```

<img src="man/figures/Covid19StripesPerCapita-1.png" style="display: block; margin: auto;" />

Or single out countries that you are interested in

``` r
plot_covid19_stripes(
  type = "confirmed", 
  countries = c("ITA", "ESP", "FRA", "GBR", "DEU", "USA", "BRA", "MEX"),
  sort_countries = "countries"
)
```

<img src="man/figures/Covid19StripesSelCountries-1.png" style="display: block; margin: auto;" />

### Map Covid-19

Finally, as Covid-19 has become a truly world-wide pandemic, I decided
to also include a basic mapping function. `map_covid19()` allows you to
map the spread of the virus at a certain date both world-wide …

``` r
map_covid19(merged, cumulative = TRUE)
```

<img src="man/figures/MapWorldWide-1.png" style="display: block; margin: auto;" />

… or for certain regions.

``` r
map_covid19(merged, type = "confirmed", region = "Europe") 
```

<img src="man/figures/MapEurope-1.png" style="display: block; margin: auto;" />

If you have enough time (takes several minutes), you can also create an
animation to visualize the spread of the virus over time.

``` r
map_covid19(
  merged, type = "confirmed", per_capita = TRUE, dates = unique(merged$date)
)
```

<img src="man/figures/AnimatedMapWorldWide-1.gif" style="display: block; margin: auto;" />

Again, you can customize the data that you want to plot and of course
you can also modify the plot itself by using normal `ggplot` syntax.

## Shiny App

Sorry, I could not resist. The options of the `plot_covid19_spread()`
make the implementation of a shiny app a little bit to tempting to pass.
The command `shiny_covid19_spread()` starts the app. Click on the image
to be taken to the online app. You can use it to customize your
`plot_covid19_spread()` display as it allows copying the plot generating
code to the clipboard, thanks to the fine
[{rclipboard}](https://github.com/sbihorel/rclipboard) package. You can
now also customize the app by providing `plot_covid19_spread()` options
as a list to the `plot_options` parameter.

<center>

[![Screenshot of `shiny_covid19_spread()`
app](man/figures/shiny_covid19_spread.png)](https://jgassen.shinyapps.io/tidycovid19/)

</center>

As the shinyapps.io server has had some issues with exhausting
connections, you can also use this [alternative
server](https://trr266.wiwi.hu-berlin.de/shiny/tidycovid19/).

## Blog posts

  - [An intro blog
    post](https://joachim-gassen.github.io/2020/05/tidycovid19-new-data-and-doc/)
    providing a quick walk-through of the package.

  - **New**: [A blog
    post](https://joachim-gassen.github.io/2021/01/vaccination-data-an-outlier-tale/)
    on the OWID vaccination data.

  - [A blog
    post](https://joachim-gassen.github.io/2020/04/tidycovid19-new-viz-and-npi_lifting/)
    on the new visuals of the package.

  - [A blog
    post](https://joachim-gassen.github.io/2020/04/covid19-explore-your-visualier-dof/)
    on the visualizer degrees of freedom that are inherent in a plot of
    the Covid-19 spread.

  - [A blog
    post](https://joachim-gassen.github.io/2020/04/scrape-google-covid19-cmr-data/)
    on the PDF scraping of the new Google Covid-19 Community Movement
    Reports.

  - [A somewhat dated blog
    post](https://joachim-gassen.github.io/2020/04/exploring-and-benchmarking-oxford-government-response-data/)

NEWS.md
# tidycovid19 0.0.0.9000

* 2021-01-03: Added hopitalization and vaccination data from Our World in Data. This implies a renaming of the function `download_owid_testing_data()` to `download_owid_data()`

* 2020-06-07: Added some options to `download_google_trends_data()` to allow the download of Google Trends data for selected countries and a wider set of countries. Also included an option to omit the sleeping delay between Google Trends queries (usage not encouraged). 

* 2020-05-21: Changed definition of active cases

* 2020-05-10: Added ECDC Covid-19 case data

* 2020-05-10: Added OWID testing data

* 2020-05-10: Changed Google CMR data to percentage*100 for consistency

* 2020-05-10: Added additional coding examples in `example_code.R`

* 2020-05-03: Added long overdue data documentation

* 2020-05-03: Provided option to download regional and U.S. county JHU CSSE data

* 2020-05-03: Provided option to download regional and U.S. county Google CMR data

* 2020-05-03: Provided option to download regional Apple MTR data

* 2020-04-30: Incorporated the new format for OxCGRT Government Respose data

* 2020-04-26: Included `map_covid19()` to produce choropleth maps of the Covid-19 spread

* 2020-04-25: Included `plot_covid19_stripes()` to produce Covid-19 stripe visuals (good if you want to compare many countries quickly)

* 2020-04-17: Changed defaults in `plot_covid19_spread()` to have more reasonable defaults for per capita plots

* 2020-04-17: Added code to download Google's Community Mobility Report data in CSV format

* 2020-04-17: Removed Google PDF download and scraping code as Google made its data available via CSV files (Yeah!) 

* 2020-04-16: Included Apple Mobility Trends Reports data

* 2020-04-13: `*_covid19_spread()`: Added additional options for per capita displays

* 2020-04-12: Included code to scrape information provided by Google COVID-19 Community Mobility Reports

* 2020-04-10: `*_covid19_spread()`: Added option to exclude unhighlighted countries

* 2020-04-10: `shiny_covid19_spread()`: Changed API to include all plot options

* 2020-04-09: `*_covid19_spread()`: Added option for averaged daily changes

* 2020-04-05: Included Oxford NPI data

* 2020-03-31: `*_covid19_spread()`: Added option for linear Y-axis

* 2020-03-31: `shiny_covid19_spread()`: Added option to copy plot code to clipboard

* 2020-03-29: Initial in-development version on Github


inst/app/app.R
library(shiny)
library(shinyjs)
library(dplyr)
library(shinyWidgets)
library(tidycovid19)
library(rclipboard)
library(lubridate)
library(zoo)

load("shiny_data.Rda")

po <- shiny_plot_options

if(as.numeric(
  difftime(now(tzone = "Europe/Berlin"),
           po$data$timestamp[1]), units = "hours") > 24)
  po$data <- download_merged_data(cached = TRUE, silent = TRUE)

ui <- fluidPage(
  rclipboardSetup(),
  useShinyjs(),
  titlePanel("Explore the Spread of Covid-19"),
  tags$p(
    "This display is based on data",
    tags$ul(
      tags$li("from the",
              tags$a(href = "https://github.com/CSSEGISandData/COVID-19",
                     "Johns Hopkins University CSSE team"),
              "on the spread of the SARS-CoV-2 virus"),
      tags$li("from the",
              tags$a(href = "https://www.acaps.org/covid19-government-measures-dataset",
                     "ACAPS governmental measures database"),
              "and"),
      tags$li(HTML(paste0("from the ",
              tags$a(href = "https://data.worldbank.org", "World Bank"), ".")))
    ),
    "It has been inspired by the displays created by John Burn-Murdoch",
    "for the Financial Times. Read this",
    tags$a(href = "https://joachim-gassen.github.io/2020/04/covid19-explore-your-visualier-dof/", "blog post"),
    "about using the visualizations and hover over the plot with your mouse to",
    "get more information about the respective data point."
  ),
  hr(),

  sidebarLayout(
    sidebarPanel(
      radioButtons(
        "type", "Which statistics do you want to display?",
        c("Confirmed Cases" = "confirmed", "Reported deaths" = "deaths",
          "Recovered Cases" = "recovered", "Active cases" = "active"),
        po$type
      ),
      sliderInput(
        "min_cases",
        "Number of cases to mark day zero on X-axis",
        min = 0, max = max(20000, po$min_cases), step = 50, value = po$min_cases
      ),
      sliderInput(
        "min_by_ctry_obs",
        "Required number of days to include country",
        min = 0, max = max(30, po$min_by_ctry_obs), value = po$min_by_ctry_obs
      ),
      sliderInput(
        "edate_cutoff",
        "How many days do you want to display?",
        min = 10, max = max(270, po$edate_cutoff), value = po$edate_cutoff
      ),
      hr(),
      checkboxInput(
        "cumulative",
        "Deselect for daily changes instead of cumulative values",
        po$cumulative
      ),
      sliderInput(
        "change_ave",
        "For daily changes, days to average over",
        min = 1, max = max(po$change_ave, 14), value = po$change_ave
      ),
      hr(),
      checkboxInput(
        "per_capita",
        "Display relative to population (per capita)",
        po$per_capita
      ),
      checkboxInput(
        "per_capita_x_axis",
        "Construct the X axis relative to population (per capita)",
        po$per_capita_x_axis,
      ),
      helpText(
        "When checking the above, consider adjusting number of cases to",
        "reflect that cases are now counted by 100,000 inhabitants."
      ),
      sliderInput(
        "population_cutoff",
        paste("Limit sample to countries with at least .. million inhabitants",
              "(useful for per capita displays)"),
        min = 0, max = max(100, po$population_cutoff), value = po$population_cutoff
      ),
      hr(),
      checkboxInput(
        "log_scale",
        "Use logarithmic scaling for Y-axis",
        po$log_scale
      ),
      selectInput(
        "intervention",
        "Select intervention type to highlight by points",
        c("None" = "none", "General lockdowns" = "lockdown",
          "Social distancing"= 'soc_dist', "Movement restrictions" = 'mov_rest',
          "Public health measures" = 'pub_health',
          "Social and economic measures" = 'soc_econ'),
        ifelse(!is.null(po$intervention), po$intervention, "none")
      ),
      uiOutput("SelectCountriesHL"),
      checkboxInput(
        "exclude_others",
        "Check to exclude unhighlighted countries",
        po$exclude_others
      ),
      hr(),
      uiOutput("SendCodeToClipboard")
    ),
    mainPanel(div(
      style = "position:relative",
      plotOutput(
        "Covid19Plot", width = "100%", height = "600px",
        hover = hoverOpts("plot_hover", delay = 100, delayType = "debounce"))
    ),
    uiOutput("Covid19PlotHover"),
    hr(),
    HTML("<p><center>Based on the <a href=https://joachim-gassen.github.io/tidycovid19>",
         "{tidycovid19} R Package</a>, <a href=https://twitter.com/JoachimGassen>",
         "Joachim Gassen</a>,<br>",
         "<a href=https://www.wiwi.hu-berlin.de/rewe>",
         "Humboldt-Universität zu Berlin</a> and",
         "<a href=https://www.accounting-for-transparency.de>",
         "TRR 266 Accounting for Transparency</a>, 2020</center></p>")
    ))
)

server <- function(session, input, output) {

  dyn_data <- reactive({
    data <- po$data %>%
      filter(population > 1e6*input$population_cutoff) %>%
      mutate(active = confirmed - recovered)

    if (!input$cumulative)
      data <- data %>%
        group_by(iso3c) %>%
        mutate(
          delta = !! sym(input$type) - dplyr::lag(!! sym(input$type)),
          change = rollmean(delta, input$change_ave, na.pad=TRUE, align="right")
        ) %>%
        ungroup()

    if(input$per_capita_x_axis) {
      data <- data %>%
        group_by(iso3c) %>%
        filter((!! sym(input$type)*1e5)/population >= input$min_cases)
    } else {
      data <- data %>%
        group_by(iso3c) %>%
        filter(!! sym(input$type) >= input$min_cases)
    }

    data %>%
      mutate(edate = as.numeric(date - min(date))) %>%
      filter(!is.na(edate)) %>%
      group_by(iso3c) %>%
      filter(n() >= input$min_by_ctry_obs) %>%
      ungroup() %>%
      filter(edate <= input$edate_cutoff) -> df_temp

    if (!input$cumulative) df_temp <- df_temp %>%
      mutate(!! input$type := change) %>%
      filter(!is.na(!! sym(input$type)))

    if (input$per_capita) df_temp <- df_temp %>%
      mutate(!! input$type := 1e5*(!! sym(input$type))/population) %>%
      filter(!is.na(!! sym(input$type)))

    return(df_temp)
  })

  ctry_selected <- reactiveVal(
    if(!is.null(po$highlight)) {
      po$data %>%
          select(country, iso3c) %>%
          unique() %>%
          filter(iso3c %in% po$highlight) -> ctries_temp
        ctries <- ctries_temp$iso3c
        names(ctries) <- ctries_temp$country
        ctries <- ctries[order(ctries_temp$country)]
        ctries
    } else "all"
  )

  ctry_list <- reactive({
    dyn_data() %>%
      select(iso3c, country) %>%
      unique() -> ctries_temp
    ctries <- ctries_temp$iso3c
    names(ctries) <- ctries_temp$country
    ctries <- ctries[order(ctries_temp$country)]
    if (isolate(ctry_selected()[1]) == "all") ctry_selected(ctries)
    else {
      still_selectable_ctries <- isolate(ctry_selected()[ctry_selected() %in% ctries])
      if (length(still_selectable_ctries) == 0) still_selectable_ctries <- ""
      ctry_selected(still_selectable_ctries)
    }
    return(ctries)
  })

  plot_code <- reactive({
     paste(
      sep = "\n",
      '# Code generated by shiny_covid19_spread() of the {tidycovid19} package',
      '# See: https://github.com/joachim-gassen/tidycovid19',
      '# Run in R/Rstudio. See https://www.r-project.org and https://www.rstudio.com',
      '# Uncomment the following to install the {tidycovid19} package',
      '',
      '# remotes::install_github("joachim-gassen/tidycovid19)',
      '',
      'library(tidycovid19)',
      '',
      'plot_covid19_spread(',
      sprintf('  type = "%s", min_cases = %d, min_by_ctry_obs = %d,',
              input$type, input$min_cases, input$min_by_ctry_obs),
      sprintf('  per_capita = %s, per_capita_x_axis = %s, population_cutoff = %d, ',
              as.character(input$per_capita),
              as.character({if (input$per_capita) input$per_capita_x_axis else FALSE}),
              input$population_cutoff),
      sprintf('  log_scale = %s, cumulative = %s, change_ave = %d,',
              as.character(input$log_scale), as.character(input$cumulative),
              input$change_ave),
      sprintf('  edate_cutoff = %d', input$edate_cutoff),
      paste0(strwrap(sprintf('  highlight = %s,',
              paste0(capture.output(dput(unname(input$highlight))), collapse = "")), 66),
              collapse = "\n  "),
      sprintf('  exclude_others = %s, intervention = %s',
              as.character(input$log_scale), ifelse(input$intervention == "none", "NULL",
                     paste0('"', input$intervention, '"'))),
      ')'
    )
  })

  observeEvent(input$cumulative, {
    if (input$cumulative) shinyjs::disable("change_ave")
    else shinyjs::enable("change_ave")
  })

  observeEvent(input$per_capita, {
    if (input$per_capita) shinyjs::enable("per_capita_x_axis")
    else {
      updateCheckboxInput(session, "per_capita_x_axis", value = FALSE)
      shinyjs::disable("per_capita_x_axis")
    }
  })
  
  observeEvent(input$per_capita_x_axis, {
    if (input$per_capita_x_axis) {
      updateSliderInput(session, "min_cases", max = 100, step = 1, value = 5)
    } else {
      updateSliderInput(
        session, "min_cases", 
        max = max(20000, po$min_cases), step = 50, value = po$min_cases
      )
    }
  })

  observeEvent(input$highlight, ignoreNULL = FALSE, ignoreInit = TRUE, {
    ctry_sel <- isolate(ctry_selected())
    if (length(input$highlight) != length(ctry_sel) ||
        !all(input$highlight %in% ctry_sel)) {
      if (!is.null(input$highlight)) {
        ctry_selected(input$highlight)
      } else ctry_selected("")
    }
  })

  output$SelectCountriesHL <- renderUI(
    pickerInput(
      inputId = "highlight",
      label = "Which countries do you want to highlight?",
      choices = ctry_list(),
      selected = isolate(ctry_selected()),
      options = list(
        `actions-box` = TRUE,
        size = 10,
        `selected-text-format` = "count > 3"
      ),
      multiple = TRUE
    )
  )

  output$SendCodeToClipboard <- renderUI({
    rclipButton("clipbtn", "Send Code for Plot to Clipboard",
                plot_code(), icon("clipboard"))
  })

  output$Covid19Plot <- renderPlot({
    req(input$min_by_ctry_obs)
    plot_covid19_spread(
      data = po$data,
      type = input$type,
      min_cases = input$min_cases,
      min_by_ctry_obs = input$min_by_ctry_obs,
      per_capita = input$per_capita,
      per_capita_x_axis = {if (input$per_capita) input$per_capita_x_axis else FALSE},
      population_cutoff = input$population_cutoff,
      log_scale = input$log_scale,
      cumulative = input$cumulative,
      change_ave = input$change_ave,
      edate_cutoff = input$edate_cutoff,
      intervention = if(input$intervention != "none") input$intervention else NULL,
      highlight = ctry_selected(),
      exclude_others = input$exclude_others
    )
  })

  output$Covid19PlotHover <- renderUI({
    req(input$plot_hover)
    hover <- input$plot_hover
    df <- dyn_data()
    if (input$exclude_others && isolate(ctry_selected()[1]) != "all" )
      df <- df %>% filter(iso3c %in% isolate(ctry_selected()))

    if (!is.null(hover)) {
      hover$mapping$x <- "edate"
      hover$mapping$y <- input$type
    }
    point <- nearPoints(df, hover, threshold = 25, maxpoints = 1, addDist = TRUE)
    if (nrow(point) == 0) return(NULL)

    # create style property for tooltip
    # background color is set so tooltip is a bit transparent
    # z-index is set so we are sure are tooltip will be on top
    style <- paste0("position:absolute; background-color: rgba(245, 245, 245, 0.85); ",
                    "left:", hover$coords_css$x + 2, "px; top:", hover$coords_css$y + 2, "px;")
    # actual tooltip created as wellPanel
    panel_text <- sprintf(
      "%s, %s:<br>%s", point$country, as.character(point$date),
      format(point[, input$type] %>% pull(), digits = 3, big.mark = ",")
    )
    if (!input$cumulative) panel_text <- paste(panel_text, "daily change in<br>")
    panel_text <- paste(panel_text, case_when(
        input$type == "confirmed" ~ "confirmed cases",
        input$type == "deaths" ~ "reported deaths",
        input$type == "recovered" ~ "recoveries",
        input$type == "active" ~ "active cases"
    ))
    if (input$per_capita)
      panel_text <- paste(panel_text, "<br>per 100,000 inhabitants")
    wellPanel(style = style, class = "well-sm", HTML(panel_text))
  })
}

shinyApp(ui = ui, server = server)
